README.md

           oooo   o8o   .o8        .oooooo..o ooooooooo.   oooooooooooo
           `888   `"'  "888       d8P'    `Y8 `888   `Y88. `888'     `8
            888  oooo   888oooo.  Y88bo.       888   .d88'  888
            888  `888   d88' `88b  `"Y8888o.   888ooo88P'   888oooo8
            888   888   888   888      `"Y88b  888          888    "
            888   888   888   888 oo     .d8P  888          888
           o888o o888o  `Y8bod8P' 8""88888P'  o888o        o888o

--------------------------------------------------------------------------------
            An ANSI C Implementation of the Sender Policy Framework
--------------------------------------------------------------------------------


  Table of Contents
  -----------------

  1.0 ............. Installation
  2.0 ............. Patches
  3.0 ............. Examples
  4.0 ............. Documentation
   4.1 ............ MTA specific
   4.2 ............ API specific
   4.3 ............ Resources
  5.0 ............. Source Code
  6.0 ............. License
  7.0 ............. Portability
  8.0 ............. Resources
   8.1 ............ Email
   8.2 ............ Developer Forums
   8.3 ............ Mailinglist
  9.0 ............. Contact
  10.0 ............ Donating


1. Installation:
----------------

./configure && make

You will find the static and dynamic libraries in the directory libtool
will have created off of the libspf src dir.:

Example: ./libspf-1_0_0/src/libpf/.libs/

You will find the libSPF Query Tool binaries (both static and dynamic) in 
a directory created by make in the libSPF root directory called 'bin'.

Example: ./libspf-1_0_0/bin/

Outside of the generic available configure options you have the following
at your disposal:


--enable-pthreads  - Default = disabled

Enable linux posix threading support which in turn switches from gethostbyname
to gethostbyname_r (if available, if not a custom reentrant wrapper is used)
and several mutexes are employed to facilitate a thread-safe environment.


--enable-spfmilter - Default = disabled

Enable the libSPF fork of Jef Poskanzer's Sendmail spfmilter.  The milter was
forked at version 0.90.  Problems with the milter are NOT to be directed to 
Jef as its not fair to ask him to support a fork when he is plenty busy with
his own.  Please forward bugs, questions, etc.. to myself (James).


--enable-spfqtool - Default = enabled

Enable the libSPF Query Tool.  spfqtool is a fully functional binary SPF 
parser which you can call from userland or even call from a CGI or your
MTA if you so wished.  Since its written in the same language as libSPF 
its very lightweight and inexpensive to call repeatedly. 

The above switches also work in reverse, so if for example you wished
to disable to building of the query tool you would do:

./configure --disable-spfqtool


--enable-debug    - Default = disabled

Enable debugging output.  For a detailed document on how debugging works
within libSPF please see the internal documenation or the online document
available at http://libSPF.org/debugging_libspf.pdf/html  By default 
debug output is pushed to /var/log/spf.log.  If you wish to change this see
the AutoConf configuration handle '--enable-debug-logfile' and disable it.


--enable-debug-logfile - Default = enabled

Enables logging debug output to /var/log/spf.log


--enable-logfile-stats - Default = disabled

Enables logfile stats to /var/log/spflog.txt which look like this:

[2005-04-18 14:24:48] result: unknown :: NievNeal3864@jlist.com [82.120.84.13], ver: 10, depth: 1, error: (policy result: (unknown) from rule (-all))
[2005-04-18 14:24:48] result: unknown :: NievNeal3864@jlist.com [82.120.84.13], ver: 10, depth: 1, error: (policy result: (unknown) from rule (-all))
[2005-04-18 17:10:49] result: softfail :: emgdxfwavxzwhk@hotmail.com [201.15.156.75], ver: 5, depth: 1, error: (policy result: (softfail) from rule (~all))
[2005-04-18 21:14:00] result: neutral :: test@spftools.infinitepenguins.net [220.117.245.29], ver: 2, depth: 1, error: (policy result: (neutral) from rule (?all))
[2005-04-18 21:14:07] result: neutral :: test@spftools.infinitepenguins.net [220.117.245.29], ver: 2, depth: 1, error: (policy result: (neutral) from rule (?all))

I've got code to do this to MySQL if anyone is interested.. I'll try to 
include it in subsequent releases to 1.0 final.


--enable-rfc-recursion - Default = disabled

Change SPF recursion to 20 (as per RFC).  This level has been deemed by members
of the SPF-DISCUSS mailing list to be excessively high, although as far as I 
know Meng still believes it to be reasonable.  I personally believe that its
excessive, at least for now, which is why I've added this option to re-enable
RFC recursion limits.


--enable-res_search - Default = disasbled

Enables DNS queries to follow default and search rules.  This turned out to cause
all sorts of problems for people who were employing default and search fules in
their resolv.conf and was causing SPF queries to return improper results.  Its 
honestly a bit of a feature if you are aware of it, and since most people never
seem to read documentation or be aware of their surroundings, this is disabled
by default.


--enable-paranoid-malloc - Default = enabled

Enable memsetting of every malloc called through the wrapper functions
through wrapper functions that handle memory allocation.


--enable-libbind - Default = disabled

Link libbind instead of libresolv


--enable-full-optimizations - Default = disabled [DEPRECATED]

Enables building of an optimized library through the employment of various GCC
compile time macros.  I currently advise against this as they are architecture
specific but in the future it is intended that they will be intelligent and
will result in the right set of macros to result in the most optimized code
generated for whatever architecture it is compiled on. [DEPRECATED]


2. Patches: 
-----------

Patches for each respective supported MTA are found within the patches
directory, followed by the name of the MTA, and then optionally should
there have been support long enough, there will be where applicable 
additional directories per MAJOR.MINOR.INCREMENTAL.  

Example: ./libspf-1_0_0/patches/Sendmail/8.13.x/

The above example contains patches against Sendmail 8.13.x (the x is literal)
and infers that a patch against any 8.13 version will be located there.


3. Examples:
------------
  
Examples follow the same layout as Patches.  The examples directory contains
example configuration files for the various MTA's supported by libSPF.

Example: ./libspf-1_0_0/examples/Sendmail/8.13.x/sendmail.cf.example


4. Documentation:
-----------------

Documentation also follows the same layout as Patches and examples.  The
Documentation directory is called 'docs' and is also found in the root of the
unpack libspf archive.  Documentation is split up into two categories.


  4.1 MTA specific documentation:
  -------------------------------

  There is a basic HOWTO document which details step-by-step how to patch and
  install a copy of the respective MTA's supported by libSPF.  The documents
  follow the following naming convention:
  
  <mta>-<mta_version>-libspf-<libspf_version>-HOWTO.[html|txt|pdf]
  
  Example: ./libspf-1_0_0/docs/qmail/net-1.05/netqmail-1.05-libspf-1.0-HOWTO.txt
  
  
  4.2 API specific documentation:
  -------------------------------
 
  API specific documentation is found in the 'API' directory inside of the
  'docs' directory.  You can point your favorite web browser to this folder
  or simply use your favorite shell or file explorer to traverse through the
  content.
  
  The primary support for learning the API can be had through the HTML doxygen
  content available within each release of libSPF starting with libSPF-1.0-RC3,
  and the most up to date version will always be available from the libSPF.org 
  website.
  
  Secondary documentation is available from within this directory structure also
  and consists of papers that will go into specific detail on a particular
  topic.
  
  Example: ./libspf-1_0_0/docs/API/debugging_libspf.pdf


  4.3 Resources:
  --------------
  
  Please see the section entitled "libSPF Resources"
  
  
5.0 Source code:
----------------

All source code can be found from within the 'src' directory off the root of
the unpacked libspf 1.0 archive.  From there each individual library or 
binary will be located in its own directory.

Example: ./libspf-1_0_0/src/libspf


6.0 License:
------------
    
libSPF is released under our own license, which is a modified version of the
Apache Software Foundation's license.  This license was chosen because we* wanted
SPF implemented as fast and as easily as possible, and it was felt that licensing 
under the GPL would inhibit use by commertial entities wishing to support
something like SPF but being unable to do so due to a restrictive license such
as the GPL.

*we (Sean Comeau and myself)

License: ./libspf-1_0_0/LICENSE 


7.0 Portability:
----------------
   
libSPF is coded in accordance with the ANSI C/89 and ANSI C/99 Standards.  The
library entirely complies with C/89, and I had to move to employ standards
from C/99 to facilitate the debugging and standard function replacements that
are implemented.  For more information please read "Debugging libSPF" which can
be found within the archive this README was enclosed in, inside of the 'docs'
directory off its root.

libSPF is easily portable to win32 although because windows has not been the
focus of interest its lagged behind more important things such as API
documentation and getting near a stable release.  If win32 portability 
interests you feel free to contact me, all thats necessary is to finish off
a patch altering the DNS functionality.

libSPF is known to 'configure', 'compile' and work on the following
architectures and the distributions associated with them:

 [+--------------------------------------------------------------------------+]
  | Architecture | Distro         | Version  | Kernel    | Compiler Version |
  [+-------------+----------------+----------+-----------+------------------+] 
   | x86         | Fedora Core    | 1        | 2.6.3     | GCC 3.3.2        |
   | x86_64      | Fedora Core    | 1        | 2.6.7     | GCC 3.3.4 #$     |
   | x86         | Fedora Core    | 2        | 2.6.7     | GCC 3.3.3/4      |
   | x86_64      | Fedora Core    | 2        | 2.6.7     | GCC 3.3.3/4 #$   |
   | x86_64      | Fedora Core    | 3        | 2.6.11    | GCC 3.4.3   #$   |
   | x86         | Gentoo         | 1.4      | 2.6.7     | GCC 3.3.3/4 @#$  |
   | x86_64      | Gentoo         | 2004.1   | 2.6.7-r11 | GCC 3.3.3/4 @#$  |
   | x86_64      | Gentoo         | 2005.1   | 2.6.10-g6 | GCC 3.4.3   @#$  | 
   | x86         | Mandrake       | 9.2      | ?         | GCC 3.3.1        |
   | x86         | Slackware      | 8.0      | 2.6.7     | GCC 2.95.3       |
   | x86         | Slackware      | 8.1      | 2.6.7     | GCC 2.95.3       |
   | x86         | Slackware      | 9.0      | 2.6.7     | GCC 3.2.2        |
   | x86         | Slackware      | 9.1      | 2.6.7     | GCC 3.3.1        |
   | x86         | Slackware      | 10.0     | 2.6.7     | GCC 3.3.4        |
   | x86         | Red Hat Linux  | 3.2.3-34 | 2.4.21-15 | GCC 3.2.3 030502 | 
  [+-------------+----------------+----------+-----------+------------------+] 
   | x86         | OpenBSD        | 3.5      | 3.5       | GCC 2.95.3 @     |
   | x86_64      | OpenBSD        | 3.5      | 3.5       | GCC 2.95.3 @     |
   | x86         | FreeBSD        | 4.4      | 4.4       | GCC 2.95.3       |
   | x86         | FreeBSD        | 4.10-PRE | 4.10-PRE  | GCC 2.95.4       | 
   | x86         | FreeBSD        | 4.11-STA | 4.11-STAB | GCC 2.95.4       |
   | x86         | FreeBSD        | 5.3      | 5.3       | GCC ??????       |
  [+-------------+----------------+----------+-----------+------------------+] 
   | PPC         | OSX/Darwin     | 10.3     | 10.3      | GCC 3.3          | 
  [+-------------+----------------+----------+-----------+------------------+] 
   | SPARC       | SunOS/Solaris  | 5.8/8.0  | 5.8/8.0   | GCC/SC 3.0.1/5.3 |
   | x86         | SunOS/Solaris  | 5.8/8.0  | 5.8/8.0   | GCC 2.95.3       | 
   | SPARC       | SunOS/Solaris  | 5.9/9.0  | 5.9/9.0   | GCC 3.3          |
   | x86         | SunOS/Solaris  | 5.9/9.0  | 5.9/9.0   | GCC/SC 3.3.2/5.5 | 
  [+-------------+----------------+----------+-----------+------------------+]
  | @ = Propolice # = ssp $ = pie | Submit: <libspf-support@codeshare.ca>    |
 [+--------------------------------------------------------------------------+]


8.0 Resources:
--------------

I've tried to provide as much support as I can through the means at my disposal.
This being said you have access to the following:

  8.1 Email:
  ----------    
      
      I'm readily available via E-Mail both on and off the list to help answer
      any questions that are not blatantly answered through a little reading.
      
      If you are a developer looking to implement and have a question, bug,
      feature, idea, etc.. DO NOT HESITATE to contact me!
      
      If you are an Administrator and you are trying to implement libSPF on
      your network, and you have exhausted the available help, DO NOT 
      HESITATE to contact me!
      
      Contact E-Mail: jcouzens@codeshare.ca
      
      
  8.2 Developer Forums:
  ---------------------
      
      The developer forums are aimed to promoting further patch work implementing
      libSPF in other MTA's and in other ways.  Its also a venue of bug-fix
      releases and feature-testing.  
      
      Developers will find this site most useful, but this doesn't exclude any 
      particular party from participating, all are welcome to be involved.
      
      Developer Foruyms: http://forums.codeshare.ca
     
  8.3 Mailing List:
  -----------------
  
      As of August 1, 2004 there is now a mailing list hosted by codeshare.ca
      entitled 'libspf'.  You can subscribe to this mailginglist quite easily
      by submitting an email to 'libspf-subscribe@codeshare.ca' with an empty
      body and the subject: 'libspf subscribe'.

      There is also an announcesments list which is 'receive only' in that 
      only libSPF developers will post to it in announcement of a new
      release or bugfix.  To subscriber to this list submit and email to
      'libspf-announce-subscribe@codeshare.ca' with an empty body and the 
      subject: 'libspf announce subscribe'.

      
9.0 Contact:
------------
    
Forums ....................... http://forums.codeshare.ca
WWW .......................... http://libspf.org
Bugs ......................... James Couzens <jcouzens@codeshare.ca>


10.0 Donations:
---------------

Honestly the best donation you can give is to provide feedback,
patches, or aide the project in one fashion or another.  If you
do wish to show your appreciation in a more tangible form, I
personally enjoy books and small electronic devices which you 
can find on my Amazon wishlist.

http://libspf.org/wishlist.html 

Link seems to break from time to time, so if it does you could
search for my name if you were so motivated, or drop me a line :)

UPDATED: Mon Apr 18 21:32:45 PDT 2005

EOF


Changes
Revision history for libspf

	- spfmilter was referenced by configure.in, but wasn't part
          of the latest tarball - discontinued?
	- applied patch by kd6lvw@yahoo.com to support SPF record
          types in addition to TXT records.
        - added default definition of T_SPF when it isn't define'd.
        - fixed a lot of minor mistakes in configure.in.
	- this version is currently (temporarily...) maintained
          by Marc Lehmann <libspf@schmorp.de>.
 



src/libspf/spf.h
/* libspf - Sender Policy Framework library
*
*  ANSI C implementation of spf-draft-200405.txt
*
*  Author: James Couzens <jcouzens@codeshare.ca>
*  Author: Sean Comeau   <scomeau@obscurity.org>
*
*  FILE: spf.h
*  DESC: main library header file
*
*  License:
*
*  The libspf Software License, Version 1.0
*
*  Copyright (c) 2004 James Couzens & Sean Comeau  All rights
*  reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions
*  are met:
*
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in
*     the documentation and/or other materials provided with the
*     distribution.
*
*  THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED
*  WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
*  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
*  DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS MAKING USE OF THIS LICESEN
*  OR ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
*  SUCH DAMAGE.
*
*/


#ifndef _SPF_H
#define _SPF_H 1

#include <sys/types.h>    /* typedefs */
#include <netinet/in.h>   /* in_addr struct fBSD */
#include <arpa/inet.h>    /* in_addr struct */

__BEGIN_DECLS

/* spf protocol version we support */
#define SPF_VERSION 1

/*
*  For reference purposes commented out are the constants based on
*  RFC 883, RFC 1034, RFC 1035.
*
*  #define PACKETSZ  512   max response packet size
*  #define MAXDNAME  1025  max uncompressed IN_TXT record
*  #define MAXCDNAME 255   max compressed IN_TXT record
*
*/

#define SPF_MAX_CNAME      5     /* we follow up max CNAMEs */
#define SPF_MAX_DEBUG      2048  /* maximum for debug code to malloc */
#define SPF_MAX_MACRO      1024  /* max length of an expanded macro */
#define SPF_MAX_STR        4096  /* max length of any dynamically allocated string */
#define SPF_MAX_DELIM      255   /* used by UTIL_count_delim() */

#define SPF_MAX_LOCAL_PART 256   /* local-part, text before @ in email addy */
#define SPF_MAX_ENV_SENDER 512   /* entire FROM: string passed by MTA */
#define SPF_MAX_CUR_DOM    256   /* text after @ in email addy for cur query */
#define SPF_MAX_UTC_TIME   22    /* time since epoch */
#define SPF_MAX_IP_ADDR    17    /* ip of remote peer - DON'T CHANGE FROM 17!! */
#define SPF_MAX_IP_VER     8     /* ip protocol version */
#define SPF_MAX_ENV_HELO   512   /* entire HELO string passed by MTA */

#define SPF_MAX_HNAME      256   /* hostname of MTA */
#define SPF_MAX_RESULT      64   /* human readable SPF result */
#define SPF_MAX_ERROR       96   /* human readable error reason */
#define SPF_MAX_EXPLAIN_S  256   /* change to EXPLAIN when cleaning up */
#define SPF_MAX_ENV_RECV   512   /* maximum length of RFC2821 header string */
#define SPF_MAX_RES_STR     12   /* maximum legnth of a res str eg: "pass" */

#define SPF_MAX_MECHANISM  256   /* maximum length of a mechanism */
#define SPF_MAX_HEADER     512   /* maximum length of header for prepend */
#define SPF_MAX_SMTP_RES   256   /* maximum length of smtp resonse string */

/* human readable string equivalents of spf responses */
#define HR_RFC2822  "Received-SPF: "
#define HR_PASS     "pass"
#define HR_NONE     "none"
#define HR_S_FAIL   "softfail"
#define HR_H_FAIL   "fail"
#define HR_ERROR    "error"
#define HR_NEUTRAL  "neutral"
#define HR_UNKNOWN  "unknown"
#define HR_UNMECH   "unknown mechanism"

/* default explanation */
#define SPF_EXPLAIN "See http://spf.pobox.com/why.html?sender=%{S}&"      \
                    "ip=%{I}&receiver=%{xR}"

/* default best guess */
#define SPF_GUESS   "v=spf1 a/24 mx/24 ptr"

/* trusted forwarder */
#define SPF_TRUSTED "v=spf1 include:spf.trusted-forwarder.org"


/* SPF_BOOL
*
*  Our own internal boolean enumeration, simple true or false.
*
*  Sendmail has issues because it makes use of SPF_FALSE and SPF_TRUE
*  Simple way around it iis to simply check and see if they are
*  defined or not since we're declaring this enumeration
*  globally.
*
*  A more graceful fix has been implemented in libspf 3.0 where
*  we now make use of SPF_SPF_TRUE and SPF_SPF_FALSE to steer clear of
*  problems such as this elsewhere.
*
*/

typedef enum SPF_BOOL
{
  SPF_FALSE = 0,
  SPF_TRUE
} SPF_BOOL;


/* SPF_RESULT
*
*  Error codes representing the result of an SPF policy examination
*
*  sucessful parse (some match was made) (+all)
*  not participating (no SPF/TXT records)
*  ~all
*  failed parse (no match made) (-all)
*  dns problem / error
*  ?all
*  permanent parsing error during record examination
*
*/
typedef enum SPF_RESULT
{
  SPF_PASS = 0,     /* + */
  SPF_NONE,
  SPF_S_FAIL,       /* ~ */
  SPF_H_FAIL,       /* - */
  SPF_ERROR,
  SPF_NEUTRAL,      /* ? */
  SPF_UNKNOWN,
  SPF_UNMECH        /* unknown mechanism */
} SPF_RESULT;


/* SPF_ACTION
*
*  Error codes representing the the action to be taken as a result
*  of the response the library was able to obtain whilst trying to
*  obtain or examin an SPF policy
*
*/
typedef enum SPF_ACTION
{
  DEFER = 0,
  TARPIT,
  ALLOW,
  REJECT
} SPF_ACTION;


/* SPF_MECHANISM
*
*  Error codes representing the various mechanism types employed
*  as defined in the RFC
*
*/
#undef VERSION /* autoconf */
typedef enum SPF_MECHANISM
{
  NO_POLICY = 0,
  VERSION,
  ALL,
  INCLUDE,
  A,
  MX,
  PTR,
  IP4,
  IP6,
  EXISTS,
  REDIRECT,
  EXPLAIN,
  DEFAULT, /* this is OLD school for early adopters = ~,?,+,- */
  UNMECH
} SPF_MECHANISM;


/* spf_result_t
*
*  Storage container used to store the result of an SPF parse.
*
*/
typedef struct spf_result_t
{
  size_t sl;      /* spf result string length */
  char s[32];     /* spf result type string */
  SPF_RESULT i;   /* spf result type */
  size_t hl;      /* length of header string */
  char h[512];    /* Received-SPF: header string */
  char p;         /* prefix identifier */
} spf_result_t;


/* policy_addr_t
*
*  Storage container used to store parsed out ip addresses in their
*  binary format (in_addr struct) and an unsigned integer containing
*  the netmask
*
*/
typedef struct policy_addr_s
{
  SPF_RESULT prefix;      /* spf prefix (-,+,~,?) */
  int8_t cidr;            /* address cidr length */
  struct in_addr addr;    /* in_addr struct (unsigned long) */
} policy_addr_t;


/* spf_config_t
*
* Global config structure
*
*/
typedef struct spf_config_s
{
  int level;    /* debug level bit */
} spf_config_t;


/* split_str_node_t
*
*  This structure is used to store where the head and tail are when
*  creating a list of split_str_node_t structures.
*
*/
typedef struct strbuf_node_s
{
  size_t                 len;    /* length of string */
  char                   *s;     /* expanded string macro */
  struct strbuf_node_s   *next;  /* pointer to next node */
} strbuf_node_t;


/* strbuf_t
*
*  This structure is used exclusively by marco.c functions and is used
*  to store macros during parsing.
*
*/
typedef struct strbuf_s
{
  strbuf_node_t   *head;      /* head node */
  u_int8_t        elements;   /* number of nodes in list */
} strbuf_t;


/* split_str_node_t
*
*  This structure is used to store where the head and tail are when
*  creating a list of split_str_node_t structures.
*
*/
typedef struct split_str_node_s
{
  size_t                   len;    /* length of string */
  char                     *s;     /* expanded string macro */
  struct split_str_node_s  *next;  /* pointer to next node */
} split_str_node_t;


/* split_str_t
*
*  This structure is used exclusively by the UTIL_reverse function and is
*  used to reverse a string using a semi-arbitrary delimiter (see
*  UTIL_is_spf_delim for valid delimiters, or the SPF RFC)
*/
typedef struct split_str_s
{
  split_str_node_t  *head;      /* head node */
  split_str_node_t  *tail;      /* tail node */
  int               elements;   /* number of nodes in list */
} split_str_t;


/* peer_info_t
*
*  Used to store information about the connected peer.  Only one of
*  SMTP protocol specific three strings will be necessarily be
*  populated in the following order of precedence: FROM, EHLO, HELO.
*
*  The ip_ver string will contain 'in-addr' if the connecting peer
*  is using IPv4, or 'ip6' if the connect
*
*  Various political and technical pressures have recently led to
*  the deprecation of the IP6.INT name space in favour of IP6.ARPA.
*  This makes IPv6 PTR data management difficult, since interim
*  clients will search IP6.INT while standard clients will search
*  IP6.ARPA. We present a simple method based on DNAME RR's
*  (see [RFC2672]) and ISC BIND9 whereby zone information can be
*  managed in a single location and then made visible in two
*  namespaces.  (http://www.isc.org/tn/isc-tn-2002-1.html)
*
*  RFC 937 (POP) states: The maximum length of a command line is 512
*  characters (including the command word and the CRLF).
*  POLICY_MATCH = SPF_TRUE
*
*  Note: from can be removed and just work on local_part@cur_dom
*
*/
typedef struct peer_info_s
{
  SPF_BOOL ALL;                           /* Was 'all' mechanism parsed */

  SPF_RESULT RES;                         /* SPF error codes for result */
  SPF_RESULT RES_P;                       /* prefix behaviour */

  SPF_BOOL use_trust;                     /* T / F trustedfwder */
  SPF_BOOL use_guess;                     /* T / F best guess */

  u_int8_t spf_ver;                       /* version of SPF */

  char *p;                                /* prefix from all mechanism */
  char *rs;                               /* ptr str result of SPF query */
  char *txt;                              /* T_TXT record from DNS */
  char *helo;                             /* HELO string */
  char *ehlo;                             /* pointer to HELO string */
  char *from;                             /* FROM string */
  char *explain;                          /* Result of an explain query */
  char *guess;                            /* Query if result is TF fails */
  char *trusted;                          /* Query if primary result is none */
  char *ptr_mhost;                        /* validate against during ptr mech */
  char *current_domain;                   /* @domain of the current query */
  char *original_domain;                  /* @domain of original query */
  char *mta_hname;                        /* ptr to MTA hname eg: mx.foo.org */
  char *r_ip;                             /* pointer to remote ip from MTA */
  char *r_vhname;                         /* validated hostname of remotehost */
  char *cur_eaddr;                        /* current email address */

  char ip_ver[SPF_MAX_IP_VER];            /* IP Protocol Version */
  char local_part[SPF_MAX_LOCAL_PART];    /* local part of address (user) */
  char utc_time[SPF_MAX_UTC_TIME];        /* The num of sec since the Epoch */
  char last_m[SPF_MAX_MECHANISM];         /* last mechanism parsed */
  char error[SPF_MAX_ERROR];              /* error (if any) cause failure */

  spf_result_t *spf_result;               /* table of str, see spf_result_t */

  struct in_addr addr;                    /* IP of the remote host (peer) */

  /*
  *  Vars below here are specific to recursion through layers of SPF queries
  *  stemming from 'include' and 'redirect' use.  In addition there is a
  *  also a buffer for CNAME records.
 */

  uint8_t spf_rlevel;                     /* recursion level */

  char *cname_buf;                        /* buf for CNAME records */
  char *redirect_buf;                     /* buf for 'redirect' instances */
  split_str_t *include_set;               /* linked list for 'include' instances */
} peer_info_t;

extern spf_config_t confg;
extern u_int8_t spf_rlevel;

/*  Main library functions (main.c) */
extern peer_info_t  *SPF_init(const char *, const char *, const char *, const char *,
                              const char *, u_int32_t, u_int32_t);
extern peer_info_t  *SPF_close(peer_info_t *);
extern SPF_RESULT   SPF_policy_main(peer_info_t *);
extern SPF_BOOL     SPF_parse_policy(peer_info_t *, const char *);

extern char         *SPF_result(peer_info_t *);
extern SPF_BOOL     SPF_smtp_from(peer_info_t *, const char *);
extern SPF_BOOL     SPF_smtp_helo(peer_info_t *, const char *);

/* Functions that alter headers (header.c) */
extern char         *SPF_build_header(peer_info_t *);
extern char         *SPF_get_explain(peer_info_t *);


__END_DECLS /* _SPF_H */

#endif /* spf.h */
